integrate swift into createTissueSpecificModel and findFluxConsistent…

mtefagh opened the pull request:
Please see https://github.com/opencobra/cobratoolbox/pull/1439

**I hereby confirm that I have:**

- [X] Tested my code on my own machine
- [X] Followed the guidelines in the [Contributing Guide](https://opencobra.github.io/cobratoolbox/docs/contributing.html)
- [X] Selected `develop` as a target branch (top left drop-down menu)

laurentheirendt commented:
@mtefagh, would you be able to integrate this PR with #1439 or do a rebase once #1439 is merged? Thanks.

mtefagh commented:
@laurentheirendt, I have merged them and also addressed the lastest comment in #1439.

laurentheirendt commented:
Great 👍  Feel free to close then #1439.

mtefagh commented:
Thanks @laurentheirendt! Done.

laurentheirendt commented:
```
The following tests failed:
testSWIFTCC.m

The reasons were as follows:
------------------------------------------------
testSWIFTCC.m:
Undefined function 'graphconncomp' for input arguments of type 'double'.
Error in partition (line 55)
   [~, C] = graphconncomp(DG, 'Directed', true);
Error in testSWIFTCC (line 37)
component = partition(model, solvers.LP{1}, 'swift');
------------------------------------------------
```
@mtefagh, there are some dependencies that may not be available on every user's computer. 

May you make sure that the test is skipped if the required toolboxes are not available? 

May you also specify this in the header of the function? It would be great actually if there was an error thrown in the function that hints the user to which toolbox is missing.

rmtfleming commented:
This function is in the Matlab bioinformatics toolbox. Do we not have it
installed in the test machines?
https://uk.mathworks.com/help/bioinfo/ref/graphconncomp.html

On Mon, 11 Mar 2019 at 07:18, Laurent Heirendt <notifications@github.com>
wrote:

> The following tests failed:
> testSWIFTCC.m
>
> The reasons were as follows:
> ------------------------------------------------
> testSWIFTCC.m:
> Undefined function 'graphconncomp' for input arguments of type 'double'.
> Error in partition (line 55)
>    [~, C] = graphconncomp(DG, 'Directed', true);
> Error in testSWIFTCC (line 37)
> component = partition(model, solvers.LP{1}, 'swift');
> ------------------------------------------------
>
> @mtefagh <https://github.com/mtefagh>, there are some dependencies that
> may not be available on every user's computer.
>
> May you make sure that the test is skipped if the required toolboxes are
> not available?
>
> May you also specify this in the header of the function? It would be great
> actually if there was an error thrown in the function that hints the user
> to which toolbox is missing.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/opencobra/cobratoolbox/pull/1440#issuecomment-471417381>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACDCOm48mWk5uiXl-vSclBNankj9hfp3ks5vVfUjgaJpZM4bdjbA>
> .
>


-- 
--
Mr. Ronan MT Fleming B.V.M.S. Dip. Math. Ph.D.
----------------------------------------------------------------------------
Assistant Professor,
Division of Systems Biomedicine and Pharmacology,
Leiden Academic Centre for Drug Research,
Faculty of Science,
Leiden University.
https://www.universiteitleiden.nl/en/staffmembers/ronan-fleming
&
H2020 Project Coordinator
Systems Medicine of Mitochondrial Parkinson’s Disease
http://sysmedpd.eu
----------------------------------------------------------------------------
Mobile:  +353 873 413 072
Skype: ronan.fleming
----------------------------------------------------------------------------
(This message is confidential and may contain privileged information. It is
intended for the named recipient only. If you receive it in error please
notify me and permanently delete the original message and any copies.)

laurentheirendt commented:
Yes, it is installed.

mtefagh commented:
@laurentheirendt, I moved that test into a separate file `testPartition.m`. Also, I wrote a note in the header of the function. Additionally, I added the assertion to the first line of the code.

laurentheirendt commented:
May you do a rebase against `develop`? Thanks.

mtefagh commented:
@laurentheirendt, done!

laurentheirendt commented:
@mtefagh, you may need to make the same changes here regarding `vecnorm`:
```
21:34:59 The following tests failed:
21:34:59 testPartition.m
21:34:59 testSWIFTCC.m
21:34:59 testSWIFTCORE.m
21:34:59 
21:34:59 The reasons were as follows:
21:34:59 ------------------------------------------------
21:34:59 testPartition.m:
21:34:59 Undefined function 'vecnorm' for input arguments of type 'double'.
21:34:59 Error in swiftcc (line 47)
21:34:59     consistent(consistent & rev == 1) = vecnorm(Z(rev(consistent) == 1, :), ...
21:34:59 Error in partition (line 81)
21:34:59             component(swiftcc(S, rev, solver)) = 1;
21:34:59 Error in partition (line 94)
21:34:59                 newcomponent(core) = partition(model, solver, algorithm);
21:34:59 Error in testPartition (line 32)
21:34:59 component = partition(model, solvers.LP{1}, 'swift');
21:34:59 ------------------------------------------------
21:34:59 ------------------------------------------------
21:34:59 testSWIFTCC.m:
21:34:59 Undefined function 'vecnorm' for input arguments of type 'double'.
21:34:59 Error in swiftcc (line 47)
21:34:59     consistent(consistent & rev == 1) = vecnorm(Z(rev(consistent) == 1, :), ...
21:34:59 Error in testSWIFTCC (line 31)
21:34:59 consistent = swiftcc(model.S, model.rev);
21:34:59 ------------------------------------------------
21:34:59 ------------------------------------------------
21:34:59 testSWIFTCORE.m:
21:34:59 Undefined function 'vecnorm' for input arguments of type 'double'.
21:34:59 Error in swiftcc (line 47)
21:34:59     consistent(consistent & rev == 1) = vecnorm(Z(rev(consistent) == 1, :), ...
21:34:59 Error in testSWIFTCORE (line 25)
21:34:59 A = swiftcc(model.S, model.rev);
21:34:59 ------------------------------------------------
```

rmtfleming commented:
Ideally it could be done with a test to see what version of matlab is being
run.

On Thu, 14 Mar 2019 at 05:55, Laurent Heirendt <notifications@github.com>
wrote:

> @mtefagh <https://github.com/mtefagh>, you may need to make the same
> changes here regarding vecnorm:
>
> 21:34:59 The following tests failed:
> 21:34:59 testPartition.m
> 21:34:59 testSWIFTCC.m
> 21:34:59 testSWIFTCORE.m
> 21:34:59
> 21:34:59 The reasons were as follows:
> 21:34:59 ------------------------------------------------
> 21:34:59 testPartition.m:
> 21:34:59 Undefined function 'vecnorm' for input arguments of type 'double'.
> 21:34:59 Error in swiftcc (line 47)
> 21:34:59     consistent(consistent & rev == 1) = vecnorm(Z(rev(consistent) == 1, :), ...
> 21:34:59 Error in partition (line 81)
> 21:34:59             component(swiftcc(S, rev, solver)) = 1;
> 21:34:59 Error in partition (line 94)
> 21:34:59                 newcomponent(core) = partition(model, solver, algorithm);
> 21:34:59 Error in testPartition (line 32)
> 21:34:59 component = partition(model, solvers.LP{1}, 'swift');
> 21:34:59 ------------------------------------------------
> 21:34:59 ------------------------------------------------
> 21:34:59 testSWIFTCC.m:
> 21:34:59 Undefined function 'vecnorm' for input arguments of type 'double'.
> 21:34:59 Error in swiftcc (line 47)
> 21:34:59     consistent(consistent & rev == 1) = vecnorm(Z(rev(consistent) == 1, :), ...
> 21:34:59 Error in testSWIFTCC (line 31)
> 21:34:59 consistent = swiftcc(model.S, model.rev);
> 21:34:59 ------------------------------------------------
> 21:34:59 ------------------------------------------------
> 21:34:59 testSWIFTCORE.m:
> 21:34:59 Undefined function 'vecnorm' for input arguments of type 'double'.
> 21:34:59 Error in swiftcc (line 47)
> 21:34:59     consistent(consistent & rev == 1) = vecnorm(Z(rev(consistent) == 1, :), ...
> 21:34:59 Error in testSWIFTCORE (line 25)
> 21:34:59 A = swiftcc(model.S, model.rev);
> 21:34:59 ------------------------------------------------
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/opencobra/cobratoolbox/pull/1440#issuecomment-472714329>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACDCOvxz3uq6PVTW5eEdh93qIUrdsnbvks5vWeRHgaJpZM4bdjbA>
> .
>


-- 
--
Mr. Ronan MT Fleming B.V.M.S. Dip. Math. Ph.D.
----------------------------------------------------------------------------
Assistant Professor,
Division of Systems Biomedicine and Pharmacology,
Leiden Academic Centre for Drug Research,
Faculty of Science,
Leiden University.
https://www.universiteitleiden.nl/en/staffmembers/ronan-fleming
&
H2020 Project Coordinator
Systems Medicine of Mitochondrial Parkinson’s Disease
http://sysmedpd.eu
----------------------------------------------------------------------------
Peer-reviewed publications: https://goo.gl/FZPG23
Mobile:  +353 873 413 072
Skype: ronan.fleming
----------------------------------------------------------------------------
(This message is confidential and may contain privileged information. It is
intended for the named recipient only. If you receive it in error please
notify me and permanently delete the original message and any copies.)

mtefagh commented:
@laurentheirendt, done! @rmtfleming, I replaced `vecnorm` with the usual matrix multiplication, so now it runs on older versions as well. There was no need to exclude some Matlab versions.

mtefagh commented:
@laurentheirendt @rmtfleming, I did try several times to solve the issues on the older versions of Matlab. As I only have R2018b, each time I committed the changes and pushed it to see the output of the console on ARTENOLIS. None of `gurobi`, `ibm_cplex`, etc. solved my problem so finally I had to replace iAF1260 model by the smaller ecoli_core_model for the test and this makes it work even on older versions.

laurentheirendt commented:
@mtefagh, may you have a look at why the builds on ARTENOLIS are failing? Thanks.

laurentheirendt commented:
Closing in favor of #1470.
